src/WebJobs.Script/Description/FunctionGroupListenerDecorator.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Listeners;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Description
+{
+    internal class FunctionGroupListenerDecorator : IListenerDecorator
+    {
+        private static readonly NoOpListener _noOpListener = new();
+        private readonly IFunctionMetadataManager _metadata;
+        private readonly IEnvironment _environment; // TODO: replace options pattern
+        private readonly ILogger _logger;
+
+        public FunctionGroupListenerDecorator(
+            IFunctionMetadataManager metadata,
+            IEnvironment environment,
+            ILogger<FunctionGroupListenerDecorator> logger)
+        {
+            _metadata = metadata ?? throw new ArgumentNullException(nameof(metadata));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public IListener Decorate(ListenerDecoratorContext context)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (!_environment.TryGetFunctionsTargetGroup(out string targetGroup))
+            {
+                // If no group configured, short-circuit.
+                return context.Listener;
+            }
+
+            _logger.LogInformation("Function group target is {targetGroup}", targetGroup);
+            string functionName = context.FunctionDefinition.Descriptor.ShortName;
+            if (!_metadata.TryGetFunctionMetadata(functionName, out FunctionMetadata m))
+            {
+                _logger.LogWarning("Unable to find function metadata for function {functionName}", functionName);
+                return context.Listener;
+            }
+
+            string group = m.GetFunctionGroup() ?? functionName;
+            if (string.Equals(targetGroup, group, StringComparison.OrdinalIgnoreCase))
```

Comment:
Just for my understanding, this is saying that if a given function is part of a function group (specifically the function group of this host instance), it is enabled for this instance of the host, otherwise disabled. How do we determine which host instance is assigned which functions group? Is that where the functions scaling and DataRole comes into play?